R/summarise_draws.R

```diff
@@ -124,23 +124,51 @@ summarise_draws.draws <- function(x, ..., .args = list()) {
   x <- repair_draws(x)
   x <- as_draws_array(x)
   variables <- variables(x)
-  out <- named_list(variables, values = list(named_list(names(funs))))
-  for (v in variables) {
-    draws <- drop_dims_or_classes(x[, , v], dims = 3, reset_class = TRUE)
-    args <- c(list(draws), .args)
-    for (m in names(funs)) {
-      out[[v]][[m]] <- do.call(funs[[m]], args)
-      if (rlang::is_named(out[[v]][[m]])) {
-        # use returned names to label columns
-        out[[v]][[m]] <- rbind(out[[v]][[m]])
-      }
+
+  # get length and output names, calculated on the first variable
+  v1 <- variables[1]
+  draws1 <- drop_dims_or_classes(x[, , v1], dims = 3)
+  args1 <- c(list(draws1), .args)
+  out1 <- named_list(names(funs))
+  for (m in names(funs)) {
+    out1[[m]] <- do.call(funs[[m]], args1)
+  }
+  the_names <- vector()
+  for (i in 1:length(out1)){
+    if (rlang::is_named(out1[[i]])) {
+      the_names <- c(the_names, names(out1[[i]]))
+    } else if (length(out1[[i]]) > 1) {
+      the_names <- c(the_names, paste0(names(out1)[i], ".", c(1:length(out1[[i]]))))
+    } else {
+      the_names <- c(the_names, names(out1)[i])
     }
-    out[[v]] <- do.call(cbind, out[[v]])
   }
-  out <- tibble::as_tibble(do.call(rbind, out))
-  if ("variable" %in% names(out)) {
+  
+  # Check for naming issues prior do doing lengthy computation
+  if ("variable" %in% the_names) {
     stop_no_call("Name 'variable' is reserved in 'summarise_draws'.")
   }
+  
+  # Pre-allocate matrix to store output
+  out <- matrix(NA, nrow = length(variables), ncol = length(unlist(out1)))
+  colnames(out) <- the_names
+  out[1,] <- unlist(out1)
+  
+  # Do the computation for all remaining variables
+  if(length(variables) > 1L){
+    for (v_ind in 2:length(variables)) {
+      v <- variables[v_ind]
+      draws <- drop_dims_or_classes(x[, , v], dims = 3)
```

Comment:
since we now use these few number of lines twice, once for the first variable and the second time for the rest, is it worth making a small helper function out of it?

Reply:
Would you prefer that the helper function be defined inside `summarise_draws.draws`, in summarise_draws.R but outside of any existing functions, or somewhere else in the codebase?

Reply:
Both, inside summarise_draws.draws, in summarise_draws.R would be fine for me.